File: 2023/python/solutions/day05.py
from functools import reduce
from itertools import zip_longest
from typing import NamedTuple

from lib.python import ProblemInput, ProblemResult, Solution


class MapEntry(NamedTuple):
    dst: int
    src: int
    len: int


Seed = int
Map = list[MapEntry]


class SeedRange(NamedTuple):
    start: int
    len: int


class Solution05(Solution):
    @staticmethod
    def results() -> list[ProblemResult]:
        return [35, 379811651, 46, 27992443]

    @staticmethod
    def solve_part01(data: ProblemInput, is_sample: bool) -> ProblemResult:
        seeds, maps = parse(data)
        return min(reduce(apply_map, maps, s) for s in seeds)

    @staticmethod
    def solve_part02(data: ProblemInput, is_sample: bool) -> ProblemResult:
        seeds, maps = parse2(data)
        return min(sr.start for s in seeds for sr in reduce(apply_map2, maps, [s]))


def parse(lines: ProblemInput) -> tuple[list[Seed], list[Map]]:
    seeds = [int(s) for s in lines[0].removeprefix("seeds: ").split()]
    maps = []
    current_map: list[MapEntry] = []
    for line in lines[2:]:
        if line.isspace():
            maps.append(current_map)
            current_map = []
            continue

        if line.endswith("map:\n"):
            continue

        d, s, l = [int(n) for n in line.split()]
        current_map.append(MapEntry(d, s, l))

    maps.append(current_map)
    return seeds, maps


def parse2(lines: ProblemInput) -> tuple[list[SeedRange], list[Map]]:
    seeds, maps = parse(lines)
    seed_ranges = [SeedRange(*pair) for pair in zip_longest(*[iter(seeds)] * 2)]  # type: ignore
    return seed_ranges, maps


def apply_map(value: Seed, m: Map) -> Seed:
    return next(
        (
            entry.dst + (value - entry.src)
            for entry in m
            if value in range(entry.src, entry.src + entry.len)
        ),
        value,
    )


def apply_map2(values: list[SeedRange], m: Map) -> list[SeedRange]:
    mapped_values = []
    remaining_values = values
    for value in remaining_values:
        for entry in m:
            entry_range = range(entry.src, entry.src + entry.len)
            init_fits = value.start in entry_range
            tail_fits = value.start + value.len - 1 in entry_range

            if init_fits and tail_fits:
                # Whole range can be mapped
                mapped_values.append(SeedRange(entry.dst + (value.start - entry.src), value.len))
                break
            elif init_fits:
                # Two ranges, start fits in current map entry
                mapped_len = (entry.src + entry.len) - value.start
                mapped_values.append(SeedRange(entry.dst + (value.start - entry.src), mapped_len))
                remaining_values.append(
                    SeedRange(value.start + mapped_len, value.len - mapped_len)
                )
                break
            elif tail_fits:
                # Two ranges, end fits in current map entry
                mapped_len = (value.start + value.len) - entry.src
                mapped_values.append(SeedRange(entry.dst, mapped_len))
                remaining_values.append(SeedRange(value.start, value.len - mapped_len))
                break
        else:
            mapped_values.append(value)

    return mapped_values

File: 2023/rust/solutions/day05.rs
use itertools::Itertools;

use aoc_lib::prelude::solution::Solution;
use aoc_lib::prelude::types::{ProblemInput, ProblemResult, ToResult};
struct MapEntry {
    dst: u64,
    src: u64,
    len: u64,
}

impl MapEntry {
    fn in_range(&self, value: u64) -> bool {
        (self.src..(self.src + self.len)).contains(&value)
    }

    fn map(&self, value: u64) -> u64 {
        self.dst + (value - self.src)
    }
}

#[derive(Clone, Copy)]
struct SeedRange {
    first: u64,
    len: u64,
}

impl SeedRange {
    fn last(&self) -> u64 {
        self.first + self.len - 1
    }
}

type Seed = u64;
type Map = Vec<MapEntry>;

pub struct Solution05;

impl Solution05 {
    fn parse(&self, input: ProblemInput) -> (Vec<Seed>, Vec<Map>) {
        let seeds = input
            .lines()
            .get(0)
            .unwrap()
            .strip_prefix("seeds: ")
            .unwrap()
            .split_whitespace()
            .map(|s| s.parse().unwrap())
            .collect_vec();

        let mut maps = vec![];
        let mut current_map = vec![];

        for line in input.lines().iter().skip(2) {
            if line.chars().all(char::is_whitespace) {
                maps.push(current_map);
                current_map = vec![];
                continue;
            }

            if line.ends_with("map:") {
                continue;
            }

            let values = line
                .split_whitespace()
                .map(|v| v.parse().unwrap())
                .take(3)
                .collect_tuple()
                .unwrap();
            let (d, s, l) = values;
            current_map.push(MapEntry { dst: d, src: s, len: l });
        }

        maps.push(current_map);
        (seeds, maps)
    }

    fn parse2(&self, lines: ProblemInput) -> (Vec<SeedRange>, Vec<Map>) {
        let (seeds, maps) = self.parse(lines);
        let seed_ranges = seeds
            .into_iter()
            .tuples()
            .map(|(s, l)| SeedRange { first: s, len: l })
            .collect();
        (seed_ranges, maps)
    }

    fn apply_map(&self, value: Seed, map: &Map) -> Seed {
        map.iter()
            .find(|entry| entry.in_range(value))
            .map_or(value, |entry| entry.map(value))
    }

    fn apply_map2(&self, values: Vec<SeedRange>, map: &Map) -> Vec<SeedRange> {
        let mut mapped_values = vec![];
        let mut remaining_values = values;

        loop {
            let mut new_values = vec![];

            for value in remaining_values.iter() {
                let mut is_mapped = false;
                for entry in map {
                    let init_fits = entry.in_range(value.first);
                    let tail_fits = entry.in_range(value.last());

                    if init_fits && tail_fits {
                        mapped_values.push(SeedRange {
                            first: entry.map(value.first),
                            len: value.len,
                        });
                        is_mapped = true;
                        break;
                    } else if init_fits {
                        let mapped_len = (entry.src + entry.len) - value.first;
                        mapped_values.push(SeedRange {
                            first: entry.map(value.first),
                            len: mapped_len,
                        });
                        new_values.push(SeedRange {
                            first: value.first + mapped_len,
                            len: value.len - mapped_len,
                        });
                        is_mapped = true;
                        break;
                    } else if tail_fits {
                        let mapped_len = (value.first + value.len) - entry.src;
                        mapped_values.push(SeedRange {
                            first: entry.dst,
                            len: mapped_len,
                        });
                        new_values.push(SeedRange {
                            first: value.first,
                            len: value.len - mapped_len,
                        });
                        is_mapped = true;
                        break;
                    }

                    // Forgot to check if middle fits
                    // (so that the middle is mapped and init and tail have to be rechecked)
                    // Apparently thats not a problem
                }

                if !is_mapped {
                    mapped_values.push(*value);
                }
            }

            if new_values.is_empty() {
                break;
            }

            remaining_values = new_values;
        }

        mapped_values
    }
}

impl Solution for Solution05 {
    fn solve_version01(&self, input: ProblemInput, _is_sample: bool) -> ProblemResult {
        let (seeds, maps) = self.parse(input);
        seeds
            .iter()
            .map(|s| maps.iter().fold(*s, |acc, m| self.apply_map(acc, m)))
            .min()
            .unwrap()
            .to_result()
    }

    fn solve_version02(&self, input: ProblemInput, _is_sample: bool) -> ProblemResult {
        let (seeds, maps) = self.parse2(input);
        seeds
            .into_iter()
            .flat_map(|s| maps.iter().fold(vec![s], |acc, m| self.apply_map2(acc, m)))
            .map(|sr| sr.first)
            .min()
            .unwrap()
            .to_result()
    }
}
